### ISSUE TITLE
Keda operator facing issue in finding offset block

### ISSUE THREAD

**Issue by sinanub18** [kedacore/keda#4215]
### Report

Getting error from keda operator saying "error":"error finding offset block for topic XXX-XXX-XXX and partition 1"

`{"level":"error","ts":"2023-02-08T16:47:13Z","logger":"kafka_scaler","msg":"","type":"ScaledObject","namespace":"XXX","name":"XXX-XXX-XXX","error":"error finding offset block for topic XXX.XXX-XXX and partition 1","stacktrace":"github.com/kedacore/keda/v2/pkg/scalers.(*kafkaScaler).getLagForPartition\n\t/workspace/pkg/scalers/kafka_scaler.go:448\ngithub.com/kedacore/keda/v2/pkg/scalers.(*kafkaScaler).getTotalLag\n\t/workspace/pkg/scalers/kafka_scaler.go:597\ngithub.com/kedacore/keda/v2/pkg/scalers.(*kafkaScaler).GetMetricsAndActivity\n\t/workspace/pkg/scalers/kafka_scaler.go:568\ngithub.com/kedacore/keda/v2/pkg/scaling/cache.(*ScalersCache).GetMetricsForScaler\n\t/workspace/pkg/scaling/cache/scalers_cache.go:77\ngithub.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).GetScaledObjectMetrics\n\t/workspace/pkg/scaling/scale_handler.go:439\ngithub.com/kedacore/keda/v2/pkg/metricsservice.(*GrpcServer).GetMetrics\n\t/workspace/pkg/metricsservice/server.go:45\ngithub.com/kedacore/keda/v2/pkg/metricsservice/api._MetricsService_GetMetrics_Handler\n\t/workspace/pkg/metricsservice/api/metrics_grpc.pb.go:79\ngoogle.golang.org/grpc.(*Server).processUnaryRPC\n\t/workspace/vendor/google.golang.org/grpc/server.go:1340\ngoogle.golang.org/grpc.(*Server).handleStream\n\t/workspace/vendor/google.golang.org/grpc/server.go:1713\ngoogle.golang.org/grpc.(*Server).serveStreams.func1.2\n\t/workspace/vendor/google.golang.org/grpc/server.go:965"}`

### Expected Behavior

Keda operator should face issue to find offset block

### Actual Behavior

Throws error with error finding offset block for topic 

### Steps to Reproduce the Problem

1. Deploy KEDA 2.9.0
2. Check logs of keda operator

### Logs from KEDA operator

`{"level":"error","ts":"2023-02-08T16:47:13Z","logger":"kafka_scaler","msg":"","type":"ScaledObject","namespace":"YYY","name":"XXX-XXX,"error":"error finding offset block for topic XX.XXX-XXX and partition 2","stacktrace":"github.com/kedacore/keda/v2/pkg/scalers.(*kafkaScaler).getLagForPartition\n\t/workspace/pkg/scalers/kafka_scaler.go:448\ngithub.com/kedacore/keda/v2/pkg/scalers.(*kafkaScaler).getTotalLag\n\t/workspace/pkg/scalers/kafka_scaler.go:597\ngithub.com/kedacore/keda/v2/pkg/scalers.(*kafkaScaler).GetMetricsAndActivity\n\t/workspace/pkg/scalers/kafka_scaler.go:568\ngithub.com/kedacore/keda/v2/pkg/scaling/cache.(*ScalersCache).GetMetricsForScaler\n\t/workspace/pkg/scaling/cache/scalers_cache.go:77\ngithub.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).GetScaledObjectMetrics\n\t/workspace/pkg/scaling/scale_handler.go:439\ngithub.com/kedacore/keda/v2/pkg/metricsservice.(*GrpcServer).GetMetrics\n\t/workspace/pkg/metricsservice/server.go:45\ngithub.com/kedacore/keda/v2/pkg/metricsservice/api._MetricsService_GetMetrics_Handler\n\t/workspace/pkg/metricsservice/api/metrics_grpc.pb.go:79\ngoogle.golang.org/grpc.(*Server).processUnaryRPC\n\t/workspace/vendor/google.golang.org/grpc/server.go:1340\ngoogle.golang.org/grpc.(*Server).handleStream\n\t/workspace/vendor/google.golang.org/grpc/server.go:1713\ngoogle.golang.org/grpc.(*Server).serveStreams.func1.2\n\t/workspace/vendor/google.golang.org/grpc/server.go:965"}`

### KEDA Version

2.9.0

### Kubernetes Version

1.25

### Platform

Google Cloud

### Scaler Details

Kafka Scaler

**Comment by JorTurFer** [comment]
Any idea about this @zroubalik ?

**Comment by sinanub18** [comment]
> Try to check this setting: https://keda.sh/docs/2.9/scalers/apache-kafka/#new-consumers-and-offset-reset-policy

The ScaledObjets already have the variable `offsetResetPolicy: latest` for kafka scalers

**Comment by jorgenfries** [comment]
Experiencing the same issue, and log errors.
Workaround is restarting keda operator pod :-/
Status on the specific scaledobjects are fine - no errors are logged here.

**Comment by dttung2905** [comment]
@jorgenfries could you help to try with keda 2.10 if possible ? This version will have better error logging from PR #4233. It won't directly solve your problem but might provide more information for troubleshooting at least :pray:

**Comment by jorgenfries** [comment]
@dttung2905 
Updated this weekend. _helm.sh/chart=keda-2.10.0_
Will post the output when/if it happens again.

**Comment by jorgenfries** [comment]
@johnnytardin - Actually I haven't seen it since. However I only noticed it once with the earlier version.
I am performing some intensive testing on our Kafka workloads the coming month, so if it's still present I should encounter it again soonish.

**Comment by johnnytardin** [comment]
I performed the update last week to 2.10.0 and continue to get the errors.
Here is the log but without further information.

`2023-03-27T15:28:37-03:00	2023-03-27T18:28:37Z	ERROR	kafka_scaler		{"type": "ScaledObject", "namespace": "c205ef32-a362-4405-b19f-441e3c453f89", "name": "1577f6da-4ef3-40e4-8a19-6bd1042011be", "error": "error finding offset block for topic [HIDDEN] and partition 13 from offset block: map[]"}
2023-03-27T15:28:36-03:00	2023-03-27T18:28:36Z	ERROR	kafka_scaler		{"type": "ScaledObject", "namespace": "c205ef32-a362-4405-b19f-441e3c453f89", "name": "625a973a-4c3f-4876-9740-53c01b5269d0", "error": "error finding offset block for topic [HIDDEN] and partition 0 from offset block: map[]"}
2023-03-27T15:28:33-03:00	2023-03-27T18:28:33Z	ERROR	kafka_scaler		{"type": "ScaledObject", "namespace": "c205ef32-a362-4405-b19f-441e3c453f89", "name": "41d383e6-6ba4-4bb9-93c0-b010b048c0f3", "error": "error finding offset block for topic [HIDDEN] 4 from offset block: map[]"}
2023-03-27T15:28:29-03:00	2023-03-27T18:28:29Z	ERROR	kafka_scaler		{"type": "ScaledObject", "namespace": "c205ef32-a362-4405-b19f-441e3c453f89", "name": "9c6006ab-8c93-4bb8-a973-1cf0658f87da", "error": "error finding offset block for topic [HIDDEN] 1 from offset block: map[]"}`

**Comment by jorgenfries** [comment]
I experienced it again also with 2.10.0 - same log entries as @johnnytardin 
Sadly I recreated the pod without getting the logs out before doing so -.-
However a pod restart of keda-operator fixed it again, and the scaling started correctly again.

**Comment by dttung2905** [comment]
```
"error finding offset block for topic [HIDDEN] and partition 13 from offset block: map[]"
```
Its quite weird actually. From your log (which was added since 2.10 ), the offset block map is empty which could be eventually traced back to this method call 
https://github.com/kedacore/keda/blob/ee28bf69389bca7bf6d5edd83e2bc3940254114c/pkg/scalers/kafka_scaler.go#L469-L479
which uses `ListConsumerGroupOffsets()` from the samara library. Might have something to do with samara itself idk. I'm not too sure as @jorgenfries said the error went away after pod recreation :thinking: ( which makes it even harder to reproduce consistantly ). Do you have any other ideas @JorTurFer @zroubalik ?

**Comment by sansmoraxz** [comment]
Not sure if entirely related but I was facing something similar #4466
Except that it wasn't logging any errors.

Restarting the keda-operator at least seemed to fix it (temporarily).

**Comment by oshmoun** [comment]
just had the issue here as well, here is the stacktrace from the logs
```
2023-04-28T09:31:42Z	ERROR	kafka_scaler		{"type": "ScaledObject", "namespace": "HIDDEN", "name": "HIDDEN", "error": "error finding offset block for topic HIDDEN and partition 4"}
github.com/kedacore/keda/v2/pkg/scalers.(*kafkaScaler).getLagForPartition
	/workspace/pkg/scalers/kafka_scaler.go:448
github.com/kedacore/keda/v2/pkg/scalers.(*kafkaScaler).getTotalLag
	/workspace/pkg/scalers/kafka_scaler.go:597
github.com/kedacore/keda/v2/pkg/scalers.(*kafkaScaler).GetMetricsAndActivity
	/workspace/pkg/scalers/kafka_scaler.go:568
```

I'm on version 2.9.3

**Comment by JorTurFer** [comment]
> I did not need to restart keda to fix the issue, but simply disable idle scaling by removing idleReplicaCount from the configured ScaledObjects

`idleReplicaCount ` doesn't have impact at scaler level, it's something external for KEDA itself, not for kafka trigger, I think that your problem has gone because modifying the ScaledObject you have triggered a rebuild of the scaler

**Comment by JorTurFer** [comment]
Could it be related with some internal change in sarama client? AFAIR, there isn't any change in kafka scaler related with that code?
@zroubalik ?

**Comment by sinanub18** [comment]
> Could it be related with some internal change in sarama client? AFAIR, there isn't any change in kafka scaler related with that code? @zroubalik ?

Hi @JorTurFer Its not about recent changes as I am facing this issue since 2.6.0

**Comment by oshmoun** [comment]
Would like to also mention that I have tested fallback scaling, hoping that keda would at least respect that in the case of this scaling failure, but it does not. Fallback scaling is rendered broken when this issue occurs.

**Comment by oshmoun** [comment]
> Do you have a way to reproduce this error?

unfortunately not, as it happens randomly
though if you are referring to the fallback scaling failure, I guess one can simply introduce an error in the code where the current error is reported, and check from there
```
github.com/kedacore/keda/v2/pkg/scalers.(*kafkaScaler).getLagForPartition
	/workspace/pkg/scalers/kafka_scaler.go:448
```

**Comment by JorTurFer** [comment]
> though if you are referring to the fallback scaling failure

No no, I meant the root cause

**Comment by oshmoun** [comment]
sorry for the lack of update on this matter. I have updated keda to 2.11.1, and since then the issue has not occurred. Hopefully not just a fluke, since the issue is of random nature.
But so far so good :crossed_fingers:

**Comment by leoeareis** [comment]
Any update related to that @oshmoun in new version?

**Comment by JorTurFer** [comment]
As it looks as solved, I'm going to close the issue. If the problem happens again, comment here and I can reopen it.
Just to announce it, we have added another "flavour" of kafka scaler as experimental scaler: https://keda.sh/docs/2.12/scalers/apache-kafka-go/
It's experimental because it doesn't have all the functionality yet, but it uses `kafka-go` instead of sarama client. It's covered by e2e test cases and even though there could be a missing test, the majority of them are the same (I mean, it's not just random code placed there). You can give a try too if the issue persists

**Comment by rxg8255** [comment]
Facing same issue, K8s version 1.25 and KEDA 2.9.3 
Tried restarting the KEDA operator but no luck.
Could any please let me know = the fix?

**Comment by sansmoraxz** [comment]
@rxg8255 can you update your KEDA operator to newer versions viz 2.12.0 and check if it still gives the same issue?

Alternatively, have a look at [Apache Kafka - (Experimental)](https://keda.sh/docs/2.12/scalers/apache-kafka-go/) and if your use case is supported check if issue persists.

**Comment by sansmoraxz** [comment]
Maybe try resetting your consumer group and/ or topic. Not sure if it would help.

I think there was a bug in the underlying dependency solved in 2.11 as @oshmoun stated above. https://github.com/kedacore/keda/issues/4215#issuecomment-1658858759
